content/posts/integration-test-code-coverage-in-go/index.md

```diff
@@ -0,0 +1,215 @@
+
+The steps we ended up taking to find the final percent was to  
+
+- compile services with `go test -c`,
+- add an http killswitch
+- start test services & test them
+- stop the services by calling the http killswitch endpoint
+- collect coverage reports from containers
+- merge them with gocovmerge.
+- get a shiny code coverage percentage.
+
+
+## Part I: Compiling services with go test instrumentation
+
+We want to run our service with `go test` so we can use it's `cover` flags to enable code coverage output.
+
+To do this, we need to first create a test function that works like our main, so that we can use `go test` to run our server.
+
+That is to say, take:
+
+```go
+func main() {
+    // start the service
+}
+```
+
+and make it:
+
+
+```go
+// main.go
+
+func main() {
+    runService()
+}
+
+func runService() {
+  // start the service.
+}
+```
+
+```go
+// main_test.go
+
+func TestRun(t *testing.T) {
+    runService()
+}
+```
+
+For the coverage to be output, TestRun function needs to finish. We can't just kill the process, or go test will not output a coverage profile.
+
+So we can set up an HTTP server, because it's really easy in Go. It will listen for a request which is our signal to let the function complete. Later, we can run curl to tell our services to stop.
+
+Our simple HTTP server calls a context's CancelFunc when it receives a request.
+
+```go
+type killServer struct {
+	server http.Server
+	cancel context.CancelFunc
+}
+
+func newKillServer(addr string, cancel context.CancelFunc) *killServer {
+	return &killServer{
+		server: http.Server{
+			Addr: addr,
+		},
+		cancel: cancel,
+	}
+}
+
+func (s *killServer) Start() {
+	s.server.Handler = s
+
+	fmt.Println("Started KillServer")
+
+	err := s.server.ListenAndServe()
+
+	if err == http.ErrServerClosed {
+		fmt.Println("KillServer Closed")
+	} else {
+		fmt.Println("KillServer Error:", err)
+	}
+}
+
+func (s *killServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+
+  // cancel the context
+	s.cancel()
+}
+
+```
+
+And our TestRun uses the context to exit once the context is canceled.
+
+```go
+func TestRun(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+  
+	killServer := newKillServer(":19999", cancel)
+	go killServer.Start()  
+	go runService(ctx)
+
+	<-ctx.Done()
+
+	killServer.server.Shutdown(context.Background())
+}
+```
+
+Next, we can compile the service with: `go test -c ./ -cover -covermode=count -coverpkg=./...` to create a `.test` binary, and run it with:
+
+```bash
+./my-service.test -test.coverprofile my-service.out
+# server started...
+```
+
+Now our server is running as it normally would, but with go's test tooling built in. All of the code it runs is counted for code coverage now.
+
+In another terminal, we run some tests and then kill it with the HTTP signal.
+
+```bash
+./run-tests.sh
```

Comment:
We better explain what is the content of this `run-tests.sh` file. Maybe something like: this file contains the commands to run all the cases of your test suite. It can be one or many `go test ....` commands.